src/Compilers/CSharp/Portable/Binder/Binder_Await.cs

```diff
@@ -303,61 +305,180 @@ private bool GetAwaitableExpressionInfo(
+                //    2. If `Mi` takes more than 1 parameter (named `P`), it is removed.
+                //    3. If `Mi` has a generic arity of 0, all of the following must be true, or `Mi` is removed:
+                //       1. The return type is `System.Void`
+                //       2. There is an identity or implicit reference conversion from `E` to the type of `P`.
+                //    4. Otherwise, if `Mi` has a generic arity of 1 with type param `Tm`, all of the following must be true, or `Mi` is removed:
+                //      2. The generic parameter of `E` is `Te`
+                //      3. `Ti` satisfies any constraints on `Tm`
+                //      4. `Mie` is `Mi` with `Te` substituted for `Tm`, and `Pe` is the resulting parameter of `Mie`
+                //      5. There is an identity or implicit reference conversion from `E` to the type of `Pe`
+                // 6. If only one `Mi` remains, that method is used for the following rewrites. Otherwise, we instead move to [await any other type].
+                runtimeAwaitCall = null;
+                placeholder = null;
+
+                if (expression.Type is not NamedTypeSymbol { Arity: 0 or 1 } exprType)
                 {
-                    awaitCall = SpecialMember.System_Runtime_CompilerServices_AsyncHelpers__AwaitValueTaskT_T;
-                    resultType = ((NamedTypeSymbol)expression.Type).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0];
+                    return false;
                 }
-                else
+
+                var asyncHelpersType = GetSpecialType(InternalSpecialType.System_Runtime_CompilerServices_AsyncHelpers, diagnostics, expression.Syntax);
+                if (asyncHelpersType.IsErrorType())
                 {
-                    runtimeAwaitHelper = null;
                     return false;
                 }
 
-                runtimeAwaitHelper = (MethodSymbol)GetSpecialTypeMember(awaitCall, diagnostics, expression.Syntax);
+                var awaitMembers = asyncHelpersType.GetMembers("Await");
 
-                if (runtimeAwaitHelper is null)
+                foreach (var member in awaitMembers)
                 {
-                    return false;
+                    if (!isApplicableMethod(exprType, member, node, diagnostics, this, out MethodSymbol? method, out Conversion argumentConversion))
+                    {
+                        continue;
+                    }
+
+                    if (runtimeAwaitCall is not null)
+                    {
+                        runtimeAwaitCall = null;
+                        placeholder = null;
+                        return false;
+                    }
+
+                    placeholder = new BoundAwaitableValuePlaceholder(expression.Syntax, expression.Type);
+
+                    BoundExpression argument = placeholder;
+                    if (!argumentConversion.IsIdentity)
+                    {
+                        argument = new BoundConversion(
```

Comment:
Should we use `CreateConversion` here? It has additional logic that just constructing a `BoundConversion` even for simple identity conversions.

Reply:
Yeah, I forgot we had a helper in the binder for that. Thanks for pointing it out.

Reply:
Consider adding a test that observes the difference

Reply:
I don't think there is one.